### PR TITLE
BTFS-1272: Support removing/unpin files after contracts expire - Pers…

### PR DESCRIPTION
**Pull request by steveyeom**
…istence Layer

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
BTFS-1172

* **What is the current behavior?** (You can also link to an open issue here)
No such functionality to remove files after contracts expire

* **What is the new behavior?** (You can also refer to a JIRA ticket here)
Now GC cleans after contracts expire

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
N/A

* **What dependencies / modules need to be updated as pre-requisites?** (Include relevant PR links here)
https://github.com/TRON-US/go-cid/pull/1
https://github.com/TRON-US/interface-go-btfs-core/pull/19
* **Description of changes**
Please look at the BTFS-1272 and 1099

---

* **Please check if the PR fulfills these requirements**
- [ x] The subject of this PR contains a JIRA ticket BTFS-xxxx (N/A for community PRs)
- [ x] PR description and commit messages are [good and meaningful](https://chris.beams.io/posts/git-commit/)
- [ x] Unit tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

* **Please make sure the following procedures have been applied before requesting reviewers**
- [ x] Code changes closely follow [Go Code Review Comments](https://github.com/golang/go/wiki/CodeReviewComments)
- [ x] Code has been *rebased* against latest master (`git merge` not recommended, unless you know what you are doing)
- [x ] Code changes have run through `go fmt`
- [ x] Code changes have run through `go mod tidy`
- [x ] All unit tests passed locally (`make test_go_test`)

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -110,7 +110,7 @@ require (
 	github.com/multiformats/go-multiaddr-dns v0.0.3
 	github.com/multiformats/go-multiaddr-net v0.0.1
 	github.com/multiformats/go-multibase v0.0.1
-	github.com/multiformats/go-multihash v0.0.5
+	github.com/multiformats/go-multihash v0.0.10
```
Comment (taiyangc) [inline go.mod]: It's too new, you need to base TRON-US/go-cid it on ipfs/go-cid v0.0.2

File: core/coreunix/add.go [hunk 2]
```diff
@@ -5,12 +5,11 @@ import (
 	"encoding/json"
 	"errors"
 	"fmt"
+	"github.com/TRON-US/go-btfs/pin"
 	"io"
 	gopath "path"
 	"strconv"
 
-	"github.com/TRON-US/go-btfs/pin"
```
Comment (taiyangc) [inline core/coreunix/add.go]: should stay here

File: pin/gc/gc.go [hunk 3]
```diff
@@ -148,13 +154,15 @@ func GC(ctx context.Context, bs bstore.GCBlockstore, dstor dstore.Datastore, pn
 // Descendants recursively finds all the descendants of the given roots and
 // adds them to the given cid.Set, using the provided dag.GetLinks function
 // to walk the tree.
-func Descendants(ctx context.Context, getLinks dag.GetLinks, set *cid.Set, roots []cid.Cid) error {
+func Descendants(ctx context.Context, pn pin.Pinner, getLinks dag.GetLinks, set *cid.Set, pinMap *cid.Map, pinMap2 *cid.Map, roots []cid.Cid) error {
```
Comment (taiyangc) [inline pin/gc/gc.go]: can you rename `pinMap` and `pinMap2`? The names are not descriptive.

File: pin/pin.go [hunk 4]
```diff
@@ -599,17 +679,53 @@ func (p *pinner) InternalPins() []cid.Cid {
...
+			return false, err
+		}
+		if has {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func IsExpiredPin(c cid.Cid, pinMap *cid.Map, pinMap2 *cid.Map) bool {
```
Comment (taiyangc) [inline pin/pin.go]: change pinMap and pinMap2 to meaningful names

File: core/commands/add.go [hunk 5]
```diff
@@ -140,6 +141,7 @@ You can now check what blocks have been created by:
 		cmds.BoolOption(encryptName, "Encrypt the file."),
 		cmds.StringOption(pubkeyName, "The public key to encrypt the file."),
 		cmds.StringOption(peerIdName, "The peer id to encrypt the file."),
+		cmds.Int64Option(pinDurationCountOptionName, "d", "Duration for which the object is pinned in days.").WithDefault(0),
```
[the remark this answers is not in the document]
Comment (steveyeom) [reply core/commands/add.go]: Yes

[the remark this answers is not in the document]
Comment (steveyeom) [reply core/commands/add.go]: Done

File: go.mod (hunk 1 above)
Comment (taiyangc) [inline go.mod]: run `go mod tidy`

File: pin/gc/gc.go (hunk 3 above)
Comment (taiyangc) [inline pin/gc/gc.go]: Still pinMap2

**Pull request merged**